proposals/0570-string-interpolation.rst

```diff
@@ -0,0 +1,643 @@
+::
+
+  error s"Expected: ${x + y}, got: ${result}"
+
+  let textExample = s"${name1} (age: ${age1}) encountered ${name2} (age: ${age2})"
+
+Proposed Change Specification
+-----------------------------
+
+This proposal introduces a new ``-XStringInterpolation`` extension that enables the following changes.
+
+Lexical Structure
+~~~~~~~~~~~~~~~~~
+
+Add ``interpolatedString`` and ``interpolatedStringMultiline`` to ``lexeme`` (not ``literal``, because they're not literals):
+
+.. code-block:: abnf
+
+  lexeme  → qvarid | qconid | qvarsym | qconsym
+          | literal | special | reservedop | reservedid
+          | interpolatedString
+          | interpolatedStringMultiline
+
+  interpolatedString → 's"' {graphic⟨'\' | '"' | '${'⟩ | space | escape | gap | '${' any⟨'}' | '"'⟩ '}'} '"'
+  interpolatedStringMultiline → 's"""' {{whitechar} interpolatedStringMultilineLine} '"""'
+  interpolatedStringMultilineLine → {graphic⟨'\' | '"""' | '${'⟩ | space | escape | gap | '${' ANY⟨'}' | '"'⟩ '}'}
+
+Also add ``$`` to ``charesc``:
+
+.. code-block:: abnf
+
+  charesc → a | b | f | n | r | t | v | \ | " | ' | & | $
+
+With ``$`` added to ``charesc``, interpolation can be avoided by escaping the dollar sign; e.g. ``s"\${foo}" == "${foo}"``.
+
+The following expressions are lex errors:
+
+* ``s"a ${"hello"} c"``
+
+  * This would lex the string ``s"a ${"``, then fail expecting a closing ``}``
+
+* ``s"a ${s"hello"} c"``
+
+  * This would lex the string ``s"a ${s"``, then fail expecting a closing ``}``
+
+Parser
+~~~~~~
+
+TODO: use sgraf's new design to avoid re-lexing + allow nested strings/braces
+
+When an interpolated string is lexed, we'll iteratively split on ``${`` and ``}`` pairs (ignoring escaped ``\${``) and re-lex the inner expressions to construct
+
+::
+
+  data HsExpr p
+    = ...
+    | HsInterpolatedString [Either FastString (LHsExpr p)]
+
+which contains the prefix of the interpolated string up to the first interpolated expression, then a list containing pairs of an interpolated expression and the subsequent string.
+
+Interpolated multiline strings will desugar to an interpolated single-line string in the lexer, the same `as usual <https://github.com/brandonchinn178/ghc-proposals/blob/master/proposals/0569-multiline-strings.rst>`_. So the parser will also get a ``HsInterpolatedString`` here (with the appropriate metadata in the extension field).
+
+The following expressions are parse errors:
+
+* ``s"a ${} b"``
+
+  * Expression is missing
+
+* ``s"a ${=} b"``
+
+  * Not an expression
+
+* ``s"a ${let x =} b"``
+
+  * Incomplete expression
+
+* ``s"a ${{b} c"``
+
+  * The second ``{`` is not a valid character to start an expression
+
+* ``s"a ${'}'} b"``
+
+  * This would try to lex the expression ``'``
+
+* ``s"foo ${x {- inline comment -} } bar"``
+
+  * This would try to lex the expression ``x {- inline comment -``
+
+* ``s"foo ${User{id = 123}} bar"``
+
+  * This would try to lex the expression ``User{id = 123``
```

Comment:
I feel like these should be allowed, they're normal expressions (that happen to use braces). Scala also supports using braces inside the interpolated expression without problems.